witness/golang/internal/witness/db_test.go

```diff
@@ -0,0 +1,182 @@
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// TestRoundTrip tests the happy paths (adding stuff and looking it up).
+func TestRoundTrip(t *testing.T) {
+	for _, test := range []struct {
+		desc      string
+		logID     string
+		c         Chkpt
+		extraRuns int
+	}{
+		{
+			desc:  "simple",
+			logID: "monkeys",
+			c: Chkpt{
+				Size: 123,
+				Raw:  []byte("bananas"),
+			},
+		}, {
+			desc:  "check no failure on inserting same data twice",
+			logID: "monkeys",
+			c: Chkpt{
+				Size: 123,
+				Raw:  []byte("bananas"),
+			},
+			extraRuns: 1,
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			ctx := context.Background()
+			db, err := sql.Open("sqlite3", ":memory:")
+			if err != nil {
+				t.Error("failed to open temporary in-memory DB", err)
+			}
+			defer db.Close()
+
+			d, err := NewDatabase(db)
+			if err != nil {
+				t.Error("failed to create DB", err)
+			}
+
+			prevChkpts := [2]uint64{0, test.c.Size}
+			for i := 0; i < test.extraRuns+1; i++ {
+				if err := d.SetCheckpoint(ctx, test.logID, prevChkpts[i], &test.c); err != nil {
+					t.Error("failed to set checkpoint", err)
+				}
+				got, err := d.GetLatest(test.logID)
+				if err != nil {
+					t.Error("failed to get latest", err)
+				}
+				if got.Size != test.c.Size || !bytes.Equal(got.Raw, test.c.Raw) {
+					t.Errorf("got != want (%x, %x)", got.Size, test.c.Size)
+				}
+			}
+		})
+	}
+}
+
+// TestUnknownKey should fail because nothing has been added.
+func TestUnknownKey(t *testing.T) {
+	db, err := sql.Open("sqlite3", ":memory:")
+	if err != nil {
+		t.Error("failed to open temporary in-memory DB", err)
+	}
+	defer db.Close()
+
+	d, err := NewDatabase(db)
+	if err != nil {
+		t.Error("failed to create DB", err)
+	}
+
+	logID := "monkeys"
+
+	r, err := d.GetLatest(logID)
+	if err == nil {
+		t.Fatalf("want error, but got none, result: %v", r)
+	}
+	if got, want := status.Code(err), codes.NotFound; got != want {
+		t.Fatalf("got error code %s, want %s", got, want)
+	}
+}
+
+// TestOutdatedChkpt should fail because caller's latest checkpoint is out of
+// date.
+func TestOutdatedChkpt(t *testing.T) {
+	ctx := context.Background()
+	db, err := sql.Open("sqlite3", ":memory:")
+	if err != nil {
+		t.Error("failed to open temporary in-memory DB", err)
+	}
+	defer db.Close()
+
+	d, err := NewDatabase(db)
+	if err != nil {
+		t.Fatalf("failed to create DB")
+	}
+
+	logID := "monkeys"
+
+	c1 := &Chkpt{
+		Size: 1,
+		Raw:  []byte("bananas"),
+	}
+	c2 := &Chkpt{
+		Size: 2,
+		Raw:  []byte("bananas"),
+	}
+	c3 := &Chkpt{
+		Size: 3,
+		Raw:  []byte("bananas"),
+	}
+
+	if err := d.SetCheckpoint(ctx, logID, 0, c1); err != nil {
+		t.Error("failed to set checkpoint", err)
+	}
+	if err := d.SetCheckpoint(ctx, logID, c1.Size, c2); err != nil {
+		t.Error("failed to set checkpoint", err)
+	}
+	err = d.SetCheckpoint(ctx, logID, c1.Size, c3)
+	if err == nil {
+		t.Fatalf("want error, but got none")
+	}
```

Comment:
Consider making this a table-driven test, similarly to how you did it in `TestRoundTrip`.